istioctl/cmd/wait.go

```diff
@@ -0,0 +1,211 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/spf13/cobra"
+	"k8s.io/apimachinery/pkg/api/meta"
+
+	"istio.io/istio/istioctl/pkg/kubernetes"
+	v2 "istio.io/istio/pilot/pkg/proxy/envoy/v2"
+)
+
+var (
+	forDistribution bool
+	targetResource  string
+	forDelete       bool
+	threshold       float32
+	timeout         time.Duration
+	resourceVersion string
+	//resourceVersions []string
+	rvLock sync.Mutex
+)
+
+const pollInterval = time.Second
+
+// waitCmd represents the wait command
+func waitCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "wait [flags] <target-resource>",
+		Short: "Wait for an Istio Resource",
+		Long: `Waits for the specified condition to be true of an istio resource.  For example:
+
+istioctl experimental wait --for-distribution virtual-service/default/bookinfo
```

Comment:
This syntax, type/namespace/name, is not like the rest of _istioctl_.

How would you feel about `wait` working like this:
- `istioctl x wait virtualservice bookinfo.default`
- `istioctl -n default wait virtualservice bookinfo`

The code for turning "virtualservice' into `virtual-service` would be new for istioctl.  Perhaps Pilot has a function, perhaps you make a map.
The code for getting the namespace is `name, ns := handlers.InferPodInfo(args[1], handlers.HandleNamespace(namespace, defaultNamespace))` (probably we should re-factor this; Istio has mostly been about pods so far).

Once you have the type, name, and namespace you put it back together again as resource type.

Then you can have `Use: "wait <type> <name>[.<namespace>"`

Reply:
I will look into this, but I think it's fairly late for 1.4... thoughts on timing?

Reply:
I agree with Ed and I think this is important to get right. Either proposed syntax looks good to me, not sure which is more common in istioctl though

Reply:
We need the usual _istioctl_ syntax for 1.4.  I have no problems with approving this under the old syntax for RC0, but for RC1 will you allow me to switch to the usual istioctl argument syntax?